code/discriminative.py

```diff
@@ -0,0 +1,127 @@
+from model import Model
+import torch
+import torch.nn as nn
+import torch.autograd as autograd
+import torch.optim as optim
+import utils
+import numpy as np
+import time
+import constants
+import random
+
+class DiscriminativeModel(nn.Module, Model):
+    
+    """
+    Contains defination for our discriminative model.
+    We train the model to learn how a sender and reciever pair usually behave.
+    We train a Logistic Regression as follows:
+    for each message m sent by s1 to r1,
+    we train (s1,r1,m) through a linear model with just 2 classes, with a true label of 1,
+    we also find a message m' which has not been sent from s1->r1. It can be sent by s1 to some other r.
+    we train(s1,r1,m') with a true label of 0.
+    We use the cross entropy loss to calculate our loss.
+    """
+    def __init__(self, epochs=10):
+        self.epochs = epochs
+        # keeps track of how many times the model has seen each email_id, either as a sender or receiver
+        self.emailid_train_freq = {}
+        super(DiscriminativeModel, self).__init__()
+        #50+50+50 (sender+recr+msg)
+        self.linear = nn.Linear(150, 2)
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+
+    def forward(self, s_id, r_id, email_emb):
+        """
+        Input are 3 autgrad variables representing sender_id and receiver_id. These ids will be used to lookup the
+        embeddings. Does a forward pass and returns the predicted email representation
```

Comment:
I think the docstring isn't correct. It doesn't return a predicted email representation per say. It will return class scores